enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/readreplica/UpstreamDatabaseStrategySelectorTest.java

```diff
@@ -0,0 +1,175 @@
+ */
+package org.neo4j.causalclustering.readreplica;
+
+import org.junit.Test;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+
+import org.neo4j.causalclustering.discovery.CoreAddresses;
+import org.neo4j.causalclustering.discovery.CoreTopology;
+import org.neo4j.causalclustering.discovery.TopologyService;
+import org.neo4j.causalclustering.identity.ClusterId;
+import org.neo4j.causalclustering.identity.MemberId;
+import org.neo4j.helpers.Service;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static org.neo4j.helpers.collection.Iterables.iterable;
+
+public class UpstreamDatabaseStrategySelectorTest
+{
+    private MemberId dummyMemberId = new MemberId( UUID.randomUUID() );
+
+    @Test
+    public void shouldReturnTheMemberIdFromFirstStrategyThatDoesNotThrowAnException() throws Exception
+    {
+        // given
+        UpstreamDatabaseSelectionStrategy badOne = mock( UpstreamDatabaseSelectionStrategy.class );
+        when( badOne.upstreamDatabase() ).thenThrow( new UpstreamDatabaseSelectionException( "bad times" ) );
+
+        UpstreamDatabaseSelectionStrategy anotherBadOne = mock( UpstreamDatabaseSelectionStrategy.class );
+        when( anotherBadOne.upstreamDatabase() )
+                .thenThrow( new UpstreamDatabaseSelectionException( "more bad times" ) );
+
+        UpstreamDatabaseSelectionStrategy goodOne = mock( UpstreamDatabaseSelectionStrategy.class );
+        MemberId theMemberId = new MemberId( UUID.randomUUID() );
+        when( goodOne.upstreamDatabase() ).thenReturn( theMemberId );
+
+        UpstreamDatabaseStrategySelector selector =
+                new UpstreamDatabaseStrategySelector( badOne, iterable( goodOne, anotherBadOne ) );
+
+        // when
+        MemberId result = selector.bestUpstreamDatabase();
+
+        // then
+        assertEquals( theMemberId, result );
+    }
+
+    @Test
+    public void shouldDefaultToRandomCoreServerAndLogIfCannotLoadClassFromConfig() throws Exception
+    {
+        // given
+        TopologyService topologyService = mock( TopologyService.class );
+        MemberId memberId = new MemberId( UUID.randomUUID() );
+        when( topologyService.coreServers() ).thenReturn( new CoreTopology( new ClusterId( UUID.randomUUID() ), false,
+                mapOf( memberId, mock( CoreAddresses.class ) ) ) );
+
+        ConnectToRandomUpstreamCoreServer defaultStrategy = new ConnectToRandomUpstreamCoreServer();
+        defaultStrategy.setDiscoveryService( topologyService );
+
+        UpstreamDatabaseStrategySelector selector = new UpstreamDatabaseStrategySelector( defaultStrategy );
+
+        // when
+        MemberId instance = selector.bestUpstreamDatabase();
+
+        // then
+        assertEquals( memberId, instance );
+    }
+
+    @Test
+    public void shouldUseSpecifiedStrategyInPreferenceToDefault() throws Exception
```

Comment:
I can't quite understand from the test why the dummy strategy would be chosen. In fact memberId is used for both.